amethyst_renderer/src/pass/skybox/interleaved.rs

```diff
@@ -0,0 +1,118 @@
+//! Skybox pass
+
+use amethyst_core::{
+    cgmath::{Matrix4, One},
+    specs::{Read, ReadStorage},
+    transform::GlobalTransform,
+};
+use gfx::pso::buffer::ElemStride;
+use glsl_layout::{mat4, Uniform};
+use std::marker::PhantomData;
+use {
+    error::Result,
+    get_camera,
+    pass::util::set_attribute_buffers,
+    pipe::{
+        pass::{Pass, PassData},
+        DepthMode, Effect, NewEffect,
+    },
+    set_vertex_args, ActiveCamera, Camera, Encoder, Factory, Mesh, Normal, PosNormTex, Position,
+    Query, Shape, TexCoord, VertexFormat,
+};
+
+use super::{SkyboxColor, FRAG_SRC, VERT_SRC};
+
+#[repr(C, align(16))]
+#[derive(Clone, Copy, Debug, Uniform)]
+pub(crate) struct VertexArgs {
+    proj: mat4,
+    view: mat4,
+    model: mat4,
+}
+
+/// Draw a simple gradient skybox
+///
+/// # Type Parameters:
+///
+/// * `V`: `VertexFormat`
+#[derive(Clone, Debug)]
+pub struct DrawSkybox<V> {
+    _marker: PhantomData<V>,
+    mesh: Option<Mesh>,
+}
+
+impl<V> DrawSkybox<V>
+where
+    V: Query<(Position, Normal, TexCoord)>,
+{
+    /// Create instance of `DrawSkybox` pass
+    pub fn new() -> Self {
+        DrawSkybox {
+            _marker: PhantomData {},
+            mesh: None,
+        }
+    }
+}
+
+impl<'a, V> PassData<'a> for DrawSkybox<V>
+where
+    V: Query<(Position, Normal, TexCoord)>,
+{
+    type Data = (
+        Option<Read<'a, ActiveCamera>>,
+        ReadStorage<'a, Camera>,
+        ReadStorage<'a, GlobalTransform>,
+        Read<'a, SkyboxColor>,
+    );
+}
+
+impl<V> Pass for DrawSkybox<V>
+where
+    V: Query<(Position, Normal, TexCoord)>,
+{
+    fn compile(&mut self, mut effect: NewEffect) -> Result<Effect> {
+        let verts = Shape::Cube.generate_vertices::<Vec<PosNormTex>>(None);
+        self.mesh = Some(Mesh::build(verts).build(&mut effect.factory)?);
+
+        effect
+            .simple(VERT_SRC, FRAG_SRC)
+            .without_back_face_culling()
+            .with_raw_constant_buffer(
+                "VertexArgs",
+                std::mem::size_of::<<VertexArgs as Uniform>::Std140>(),
+                1,
+            ).with_raw_vertex_buffer(PosNormTex::ATTRIBUTES, PosNormTex::size() as ElemStride, 0)
```

Comment:
Preferably we wouldn't actually handle the vertices this way but instead have them directly within the shader since they're always gonna be a box. The synchronization is just necessary although I don't remember if you can do shader defined vertices in OpenGL... What you definitely can do is removing the vertex format since it doesn't matter what it actually is in the end.

Reply:
Yeah, I admit that this is a bit of cargo-cult dev by starting with what other passes have done. The box only needs positions honestly.

Reply:
But a skybox isn't always a box?

Reply:
I just tested an update that allows you to specify any `amethyst::renderer::Shape` you like and the way I've currently written the shader there is no real visual distinction between the results. 

That would change if it were possible to supply a texture or material I suppose. I'm happy to experiment with that a bit but perhaps its useful to start with the most simple thing and expand into the territory of being more flexible once the work on rendy is underway?

Reply:
I still believe a skybox should be enough to cover most any cases, but I'll just trust you guys that people actually want to use different geometry :sweat_smile: 

That said, the other part of this comment is still valid and needs fixing. There's no point specifying which vertex type you can use and it just opens up for user error where they can waste memory unnecessarily.

Reply:
I'd very much like your advice regarding that. I do want to make it as simple as possible.

Reply:
By advice I mean help to understand more of how to implement this the way you suggest. I've only looked at other passes as examples so far.

Reply:
Thanks for the explanation on discord @magnonellie. I've updated the pass and have removed the type parameter for the vertex format.

Reply:
Great! Now you just need to remove it from the shaders too and we're good to merge! :+1: